source/grid/manifold.cc

```diff
@@ -87,6 +87,68 @@ get_new_point (const Quadrature<spacedim> &quad) const
   return p;
 }
 
+template <int dim, int spacedim>
+Point<spacedim>
+Manifold<dim, spacedim>::
+get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
+               const std::vector<double>           &weights) const
+{
+  const double tol = 1e-10;
+  const unsigned int n_points = surrounding_points.size();
+
+  Assert(n_points > 0,
+         ExcMessage("There should be at least one point."));
+
+  Assert(n_points == weights.size(),
+         ExcMessage("There should be as many surrounding points as weights given."));
+
+  Assert(std::abs(std::accumulate(weights.begin(), weights.end(), 0.0)-1.0) < tol,
+         ExcMessage("The weights for the individual points should sum to 1!"));
+
+  // The number of points is small (at most 26 in 3D), therefore try to minize
+  // overhead and use a very simple search of O(N^2) to find the order in which
+  // to loop over points.
+  std::vector<unsigned int> permutation(n_points);
+  std::vector<bool> found(n_points,false);
```

Comment:
I am not so sure (polite for: I do not think so :smile:). This quadrature object was never used like an actual quadrature object, it just happened to be used as a container for some points and weights. All places in the library fill this object with either:
- the vertices of some object (line, quad, or hex)
- midpoints of above object's lines and faces
- a combination of the above

That means the maximum number of points is `number_of_vertices + number_of_line_midpoints + number_of_faces = 26` for 3D. Actually most places that use this function (including `MappingQ` and `MappingManifold`) only call it with the vertices of the cell (8 in 3D), only `get_default_points_and_weights` uses the larger number.

It is true that MappingQGeneric calls this function more often when using higher order mappings, but it always hands over the same vertices of the cell as surrounding points, it just computes more new points the higher the polynomial degree.